panic.go

```diff
@@ -0,0 +1,93 @@
+package core
+
+import (
+	"fmt"
+	"sync/atomic"
+)
+
+// Recovered is a caught error from a panic call
+type Recovered interface {
+	Error() string
+	Recovered() any
+}
+
+// Recover attempts to recover from a panic
+func Recover() Recovered {
+	// revive:disable:defer
+	if rvr := recover(); rvr == nil {
+		// no panic
+		return nil
+	} else if p, ok := rvr.(Recovered); ok {
+		// pass-through
+		return p
+	} else {
+		// wrap it
+		return &panicError{rvr}
+	}
+	// revive:enable:defer
+}
+
+type panicError struct {
+	payload any
+}
+
+func (p *panicError) Error() string {
+	return fmt.Sprintf("panic: %s", p.payload)
+}
+
+func (p *panicError) Unwrap() error {
+	if err, ok := p.payload.(error); ok {
+		return err
+	}
+	return nil
+}
+
+func (p *panicError) Recovered() any {
+	return p.payload
+}
+
+// Catcher is a runner that catches panics
+type Catcher struct {
+	recovered atomic.Value
+}
+
+// Do calls a function, returning its organic error,
+// or the caught panic
+func (p *Catcher) Do(fn func() error) error {
+	if err := p.Try(fn); err != nil {
+		// natural death
+		return err
+	}
+
+	if err := p.Recovered(); err != nil {
+		// recovered panic
+		return err
+	}
+
+	// all good
+	return nil
+}
+
+// Try calls a function, returning its organic error,
+// or storing the recovered error for later consumption
+func (p *Catcher) Try(fn func() error) error {
```

Comment:
of God, oh God, Java looks back at me....

Reply:
I could make it private or change the name @karasz

Reply:
Nah. its fine.